iceoryx_utils/source/posix_wrapper/file_lock.cpp

```diff
@@ -0,0 +1,267 @@
+    });
+}
+
+cxx::expected<FileLockError> FileLock::initializeFileLock() noexcept
+{
+    if (m_name.empty())
+    {
+        return cxx::error<FileLockError>(FileLockError::NO_FILE_NAME_PROVIDED);
+    }
+    PathName_t fullPath(PATH_PREFIX + m_name + ".lock");
+    constexpr int createFileForReadWrite = O_CREAT | O_RDWR;
+    mode_t userReadWriteAccess = S_IRUSR | S_IWUSR;
+
+    auto openCall = cxx::makeSmartC(iox_open,
+                                    cxx::ReturnMode::PRE_DEFINED_ERROR_CODE,
+                                    {ERROR_CODE},
+                                    {},
+                                    fullPath.c_str(),
+                                    createFileForReadWrite,
+                                    userReadWriteAccess);
+
+    if (openCall.hasErrors())
+    {
+        return cxx::error<FileLockError>(convertErrnoToFileLockError(openCall.getErrNum()));
+    }
+    else
+    {
+        m_fd = openCall.getReturnValue();
+
+        auto lockCall = cxx::makeSmartC(
+            iox_flock, cxx::ReturnMode::PRE_DEFINED_ERROR_CODE, {ERROR_CODE}, {EWOULDBLOCK}, m_fd, LOCK_EX | LOCK_NB);
+
+        if (lockCall.hasErrors())
+        {
+            closeFileDescriptor();
+            // possible errors in closeFileDescriptor() are masked and we inform the user about the actual error
+            return cxx::error<FileLockError>(convertErrnoToFileLockError(openCall.getErrNum()));
+        }
+        else if (lockCall.getErrNum() == EWOULDBLOCK)
+        {
+            closeFileDescriptor();
+            // possible errors in closeFileDescriptor() are masked and we inform the user about the actual error
+            return cxx::error<FileLockError>(FileLockError::LOCKED_BY_OTHER_PROCESS);
+        }
+    }
+    return cxx::success<>();
+}
+
+FileLock::FileLock(FileLock&& rhs) noexcept
+{
+    *this = std::move(rhs);
+}
+
+FileLock& FileLock::operator=(FileLock&& rhs) noexcept
+{
+    if (this != &rhs)
+    {
+        if (closeFileDescriptor().has_error())
+        {
+            std::cerr << "Unable to cleanup file lock \"" << m_name
+                      << "\" in the move constructor/move assingment operator" << std::endl;
+        }
+
+        CreationPattern_t::operator=(std::move(rhs));
+
+        m_name = std::move(rhs.m_name);
+        m_fd = std::move(rhs.m_fd);
+
+        rhs.m_name.assign("");
+        rhs.m_fd = INVALID_FD;
+    }
+
+    return *this;
+}
+
+FileLock::~FileLock() noexcept
+{
+    if (closeFileDescriptor().has_error())
+    {
+        std::cerr << "unable to cleanup file lock \"" << m_name << "\" in the destructor" << std::endl;
+    }
+}
+
+cxx::expected<FileLockError> FileLock::closeFileDescriptor() noexcept
+{
+    if (isInitialized() && (m_fd != INVALID_FD))
+    {
+        auto closeCall = cxx::makeSmartC(iox_close, cxx::ReturnMode::PRE_DEFINED_ERROR_CODE, {ERROR_CODE}, {}, m_fd);
+
+        if (!closeCall.hasErrors())
+        {
+            m_fd = INVALID_FD;
```

Comment:
I think even when the close call fails we should set the file descriptor to `INVALID_FD` and `m_isInitialized` to false since the close call usually fails when either the file descriptor is already invalid or the corresponding resource is in an invalid state.

So could you please move those two lines up so that they are always called directly after the close call.

Reply:
That makes sense, done.